### PR TITLE
Schwabe/multimodule 9x

### PR DESCRIPTION
**Pull request by schwabe**
Add a few fixes for the spektrum protocol and implement UI for multimodule on 9x. I have tested the UI in the simulator and it works but the pulses are not tested, pulses probably also have the same problem as #3628. 

Therefore I will wait with adding multi as build option for 9x until someone tests the output on a 9x.

### CODE REVIEW

File: radio/src/gui/128x64/model_setup.cpp [hunk 1]
```diff
@@ -58,12 +58,19 @@ enum menuModelSetupItems {
 #if defined(CPUARM)
   ITEM_MODEL_EXTERNAL_MODULE_LABEL,
   ITEM_MODEL_EXTERNAL_MODULE_MODE,
+#ifdef MULTIMODULE
```
Comment (bsongis) [inline radio/src/gui/128x64/model_setup.cpp]: #if defined(...)

File: radio/src/gui/128x64/model_setup.cpp [hunk 2]
```diff
@@ -58,12 +58,19 @@ enum menuModelSetupItems {
...
+  ITEM_MODEL_EXTERNAL_MODULE_SUBTYPE,
+#endif
   ITEM_MODEL_EXTERNAL_MODULE_CHANNELS,
   ITEM_MODEL_EXTERNAL_MODULE_BIND,
 #if defined(PCBSKY9X) && defined(REVX)
   ITEM_MODEL_EXTERNAL_MODULE_OUTPUT_TYPE,
 #endif
   ITEM_MODEL_EXTERNAL_MODULE_FAILSAFE,
+#ifdef MULTIMODULE
```
Comment (bsongis) [inline radio/src/gui/128x64/model_setup.cpp]: same here

File: radio/src/gui/128x64/model_setup.cpp [hunk 3]
```diff
@@ -99,17 +106,41 @@ void menuModelSetup(uint8_t event)
   #define IF_EXTERNAL_MODULE_XJT(x)         (IS_MODULE_XJT(EXTERNAL_MODULE) ? (uint8_t)(x) : HIDDEN_ROW)
   #define IF_EXTERNAL_MODULE_ON(x)          (g_model.moduleData[EXTERNAL_MODULE].type == MODULE_TYPE_NONE ? HIDDEN_ROW : (uint8_t)(x))
   #define IS_D8_RX(x)                       (g_model.moduleData[x].rfProtocol == RF_PROTO_D8)
-  #define EXTERNAL_MODULE_CHANNELS_ROWS()   IF_EXTERNAL_MODULE_ON(IS_MODULE_DSM2(EXTERNAL_MODULE) ? (uint8_t)0 : (uint8_t)1)
-  #define EXTERNAL_MODULE_SETTINGS_ROWS()   (IS_MODULE_XJT(EXTERNAL_MODULE) && IS_D8_RX(EXTERNAL_MODULE)) ? (uint8_t)1 : (IS_MODULE_PPM(EXTERNAL_MODULE) || IS_MODULE_XJT(EXTERNAL_MODULE) || IS_MODULE_DSM2(EXTERNAL_MODULE)) ? (uint8_t)2 : HIDDEN_ROW
+  #define EXTERNAL_MODULE_CHANNELS_ROWS     IF_EXTERNAL_MODULE_ON((IS_MODULE_DSM2(EXTERNAL_MODULE) || IS_MODULE_MULTIMODULE(EXTERNAL_MODULE)) ? (uint8_t)0 : (uint8_t)1)
+  #define EXTERNAL_MODULE_BIND_ROWS()   (IS_MODULE_XJT(EXTERNAL_MODULE) && IS_D8_RX(EXTERNAL_MODULE)) ? (uint8_t)1 : (IS_MODULE_PPM(EXTERNAL_MODULE) || IS_MODULE_XJT(EXTERNAL_MODULE) || IS_MODULE_DSM2(EXTERNAL_MODULE) || IS_MODULE_MULTIMODULE(EXTERNAL_MODULE)) ? (uint8_t)2 : HIDDEN_ROW
```
Comment (bsongis) [inline radio/src/gui/128x64/model_setup.cpp]: alignment

File: radio/src/gui/128x64/model_setup.cpp [hunk 4]
```diff
@@ -99,17 +106,41 @@ void menuModelSetup(uint8_t event)
...
 #else
   #define OUTPUT_TYPE_ROWS() 
 #endif
   #define TRAINER_CHANNELS_ROWS()           (HIDDEN_ROW)
-  #define PORT_CHANNELS_ROWS(x)             (x==EXTERNAL_MODULE ? EXTERNAL_MODULE_CHANNELS_ROWS() : 0)
-  #define FAILSAFE_ROWS(x)                  (IS_MODULE_XJT(x) && HAS_RF_PROTOCOL_FAILSAFE(g_model.moduleData[x].rfProtocol) ? (g_model.moduleData[x].failsafeMode==FAILSAFE_CUSTOM ? (uint8_t)1 : (uint8_t)0) : HIDDEN_ROW)
-  #define CURSOR_ON_CELL                    (true)
+  #define PORT_CHANNELS_ROWS(x)             (x==EXTERNAL_MODULE ? EXTERNAL_MODULE_CHANNELS_ROWS : 0)
+
+#if defined MULTIMODULE
```
Comment (bsongis) [inline radio/src/gui/128x64/model_setup.cpp]: #if defined(...)

File: radio/src/gui/128x64/model_setup.cpp [hunk 5]
```diff
@@ -99,17 +106,41 @@ void menuModelSetup(uint8_t event)
...
+
+#if defined(MULTIMODULE)
+  #define MULTIMODULE_HAS_SUBTYPE(x)      (x == MM_RF_PROTO_FLYSKY || x == MM_RF_PROTO_FRSKY || x == MM_RF_PROTO_HISKY || x == MM_RF_PROTO_DSM2 || x == MM_RF_PROTO_YD717 || x == MM_RF_PROTO_KN || x == MM_RF_PROTO_SYMAX || x == MM_RF_PROTO_CX10 || x == MM_RF_PROTO_CG023 || x == MM_RF_PROTO_MT99XX || x == MM_RF_PROTO_MJXQ)
+  #define MULTIMODULE_MODE_ROWS(x)        (g_model.moduleData[x].multi.rfProtocol >= MM_RF_PROTO_CUSTOM ) ? (uint8_t) 3 :MULTIMODULE_HAS_SUBTYPE(g_model.moduleData[x].multi.rfProtocol) ? (uint8_t)2 : (uint8_t)1
+  #define MULTIMODULE_MODULE_ROWS         IS_MODULE_MULTIMODULE(EXTERNAL_MODULE) ? (uint8_t) 0 : HIDDEN_ROW, IS_MODULE_MULTIMODULE(EXTERNAL_MODULE) ? (uint8_t) 0 : HIDDEN_ROW,
+  #define MULTIMODULE_RFPROTO_ROWS(x)     (g_model.moduleData[x].multi.rfProtocol >= MM_RF_PROTO_CUSTOM) ? (uint8_t) 1 :MULTIMODULE_HAS_SUBTYPE(g_model.moduleData[x].multi.rfProtocol) ? (uint8_t) 0 : HIDDEN_ROW
+  #define MULTIMODULE_SUBTYPE_ROWS(x)     IS_MODULE_MULTIMODULE(x) ? MULTIMODULE_RFPROTO_ROWS(x) : HIDDEN_ROW,
+
```
Comment (bsongis) [inline radio/src/gui/128x64/model_setup.cpp]: no empty line here

File: radio/src/gui/128x64/model_setup.cpp [hunk 6]
```diff
@@ -99,17 +106,41 @@ void menuModelSetup(uint8_t event)
...
+#else
+  #define MULTIMODULE_MODE_ROWS(x)        (uint8_t)0
+  #define MULTIMODULE_MODULE_ROWS
+  #define MULTIMODULE_SUBTYPE_ROWS(x)
+#endif
+
+  #define EXTERNAL_MODULE_MODE_ROWS         (IS_MODULE_XJT(EXTERNAL_MODULE) || IS_MODULE_DSM2(EXTERNAL_MODULE) || IS_MODULE_MULTIMODULE(EXTERNAL_MODULE)) ? (uint8_t)1 : (uint8_t)0
```
Comment (bsongis) [inline radio/src/gui/128x64/model_setup.cpp]: indentation

File: radio/src/gui/128x64/model_setup.cpp [hunk 7]
```diff
@@ -99,17 +106,41 @@ void menuModelSetup(uint8_t event)
...
+
+#define CURSOR_ON_CELL                    (true)
   #define MODEL_SETUP_MAX_LINES             (1+ITEM_MODEL_SETUP_MAX)
```
Comment (bsongis) [inline radio/src/gui/128x64/model_setup.cpp]: indentation seems strange here

File: radio/src/gui/128x64/model_setup.cpp [hunk 8]
```diff
@@ -502,6 +535,12 @@ void menuModelSetup(uint8_t event)
           lcdDrawTextAtIndex(MODEL_SETUP_2ND_COLUMN+5*FW, y, STR_XJT_PROTOCOLS, 1+g_model.moduleData[EXTERNAL_MODULE].rfProtocol, menuHorizontalPosition==1 ? attr : 0);
         else if (IS_MODULE_DSM2(EXTERNAL_MODULE))
           lcdDrawTextAtIndex(MODEL_SETUP_2ND_COLUMN+5*FW, y, STR_DSM_PROTOCOLS, g_model.moduleData[EXTERNAL_MODULE].rfProtocol, menuHorizontalPosition==1 ? attr : 0);
+#if defined(MULTIMODULE)
+        else if (IS_MODULE_MULTIMODULE(EXTERNAL_MODULE)) {
+          int8_t multi_rfProto = min(g_model.moduleData[EXTERNAL_MODULE].multi.rfProtocol, (uint8_t) MM_RF_PROTO_CUSTOM);
```
Comment (bsongis) [inline radio/src/gui/128x64/model_setup.cpp]: int8_t? uint8_t?

File: radio/src/gui/128x64/model_setup.cpp [hunk 9]
```diff
@@ -518,6 +557,25 @@ void menuModelSetup(uint8_t event)
...
               if (IS_MODULE_DSM2(EXTERNAL_MODULE))
                 CHECK_INCDEC_MODELVAR(event, g_model.moduleData[EXTERNAL_MODULE].rfProtocol, DSM2_PROTO_LP45, DSM2_PROTO_DSMX);
+#if defined(MULTIMODULE)
+              else if (IS_MODULE_MULTIMODULE(EXTERNAL_MODULE)) {
+                CHECK_INCDEC_MODELVAR(event, g_model.moduleData[EXTERNAL_MODULE].multi.rfProtocol, MM_RF_PROTO_FIRST, MM_RF_PROTO_LAST);
+                if (checkIncDec_Ret) {
+                  // When in custom protocol mode the highest bit (0x20) is set to indicate the protocl we might be way above MM_RF_PROTO_LAST.
+                  // Reset to MM_RF_PROTO_LAST-1 in that case
+                  if (checkIncDec_Ret == -1 && g_model.moduleData[EXTERNAL_MODULE].multi.rfProtocol >= MM_RF_PROTO_LAST)
+                  {
```
Comment (bsongis) [inline radio/src/gui/128x64/model_setup.cpp]: no newline here

File: radio/src/gui/128x64/model_setup.cpp [hunk 10]
```diff
@@ -518,6 +557,25 @@ void menuModelSetup(uint8_t event)
...
+                    g_model.moduleData[EXTERNAL_MODULE].multi.rfProtocol = MM_RF_PROTO_LAST-1;
+                  }
+                  g_model.moduleData[EXTERNAL_MODULE].subType = 0;
+                  // Sensible default for DSM2 (same as for ppm): 6ch@11ms
+                  if (g_model.moduleData[EXTERNAL_MODULE].multi.rfProtocol == MM_RF_PROTO_DSM2)
+                    g_model.moduleData[EXTERNAL_MODULE].multi.optionValue = 6;
```
Comment (bsongis) [inline radio/src/gui/128x64/model_setup.cpp]: There is a constant somewhere: DSM2_CHANNELS. Perhaps it needs to be put in a .h in order to reuse it here

File: radio/src/gui/128x64/model_setup.cpp [hunk 11]
```diff
@@ -518,6 +557,25 @@ void menuModelSetup(uint8_t event)
...
+                  else
+                    g_model.moduleData[EXTERNAL_MODULE].multi.optionValue = 0;
+                }
+              }
+#endif
               else
                 CHECK_INCDEC_MODELVAR(event, g_model.moduleData[EXTERNAL_MODULE].rfProtocol, RF_PROTO_X16, RF_PROTO_LAST);
```
Comment (bsongis) [inline radio/src/gui/128x64/model_setup.cpp]: inside { } because the if part has

File: radio/src/gui/128x64/model_setup.cpp [hunk 12]
```diff
@@ -528,7 +586,89 @@ void menuModelSetup(uint8_t event)
         }
         break;
 #endif
-
+#if defined (MULTIMODULE)
+      case ITEM_MODEL_EXTERNAL_MODULE_SUBTYPE: {
```
Comment (bsongis) [inline radio/src/gui/128x64/model_setup.cpp]: newline here

File: radio/src/gui/128x64/model_setup.cpp [hunk 13]
```diff
@@ -528,7 +586,89 @@ void menuModelSetup(uint8_t event)
...
+        lcd_putsLeft(y, STR_SUBTYPE);
+        int8_t multi_rfProto = min(g_model.moduleData[EXTERNAL_MODULE].multi.rfProtocol, (uint8_t) MM_RF_PROTO_CUSTOM);
```
Comment (bsongis) [inline radio/src/gui/128x64/model_setup.cpp]: min<int8_t>

File: radio/src/gui/128x64/model_setup.cpp [hunk 14]
```diff
@@ -643,37 +783,73 @@ void menuModelSetup(uint8_t event)
 #endif
 
 #if defined(CPUARM)
-      case ITEM_MODEL_EXTERNAL_MODULE_FAILSAFE:
```
Comment (bsongis) [inline radio/src/gui/128x64/model_setup.cpp]: it was good here!

File: radio/src/gui/128x64/model_setup.cpp [hunk 15]
```diff
@@ -643,37 +783,73 @@ void menuModelSetup(uint8_t event)
...
             else {
-              lcdDrawFilledRect(MODEL_SETUP_2ND_COLUMN, y, LCD_W-MODEL_SETUP_2ND_COLUMN-MENUS_SCROLLBAR_WIDTH, 8);
+              lcdDrawFilledRect(MODEL_SETUP_2ND_COLUMN, y, LCD_W - MODEL_SETUP_2ND_COLUMN - MENUS_SCROLLBAR_WIDTH, 8);
             }
           }
         }
-        break;
+#if defined(MULTIMODULE)
+        else if (IS_MODULE_MULTIMODULE(moduleIdx)) {
+          switch (g_model.moduleData[moduleIdx].multi.rfProtocol) {
```
Comment (bsongis) [inline radio/src/gui/128x64/model_setup.cpp]: newline

File: radio/src/gui/gui_common_arm.cpp [hunk 16]
```diff
@@ -482,22 +482,27 @@ bool isSourceAvailableInResetSpecialFunction(int index)
...
+
 bool isModuleAvailable(int module)
 {
 #if defined(CROSSFIRE)
   if (module == MODULE_TYPE_CROSSFIRE && g_model.moduleData[INTERNAL_MODULE].rfProtocol != RF_PROTO_OFF) {
     return false;
   }
+#else
+  if (module == MODULE_TYPE_CROSSFIRE)
+    return false;
```
Comment (bsongis) [inline radio/src/gui/gui_common_arm.cpp]: here I would prefer more { }

File: radio/src/gui/gui_common_arm.cpp [hunk 17]
```diff
@@ -514,6 +519,30 @@ bool isRfProtocolAvailable(int protocol)
   return true;
 }
 
+#if defined(CPUARM)
+bool isTelemetryProtocolAvailable(int protocol)
+{
+
+#if defined(PCBTARANIS)
+  if(protocol == PROTOCOL_FRSKY_D_SECONDARY && g_eeGeneral.serial2Mode !=UART_MODE_TELEMETRY)
```
Comment (bsongis) [inline radio/src/gui/gui_common_arm.cpp]: if + space

File: radio/src/gui/128x64/model_setup.cpp (hunk 14 above)
Comment (schwabe) [inline radio/src/gui/128x64/model_setup.cpp]: Clion screwed up the  identation here. This should have only move the lcd_putsLeft after the if ```(IS_MODULE_XJT(moduleIdx)```

File: radio/src/opentx.h [hunk 18]
```diff
@@ -409,6 +409,12 @@ void memswap(void * a, void * b, uint8_t size);
   #define MAX_TRAINER_CHANNELS()            (8)
 #endif
 
+#if defined(MULTIMODULE)
+#define IS_MODULE_MULTIMODULE(idx)      (idx==EXTERNAL_MODULE && g_model.moduleData[EXTERNAL_MODULE].type==MODULE_TYPE_MULTIMODULE)
```
Comment (bsongis) [inline radio/src/opentx.h]: 2nd column could be aligned to 40 chars

File: radio/src/gui/128x64/model_setup.cpp (hunk 10 above)
Comment (schwabe) [inline radio/src/gui/128x64/model_setup.cpp]: That is only a coincidene to be 6 as well, but I will add a proper define for it.

**Comment by bsongis** [comment]
Only cosmetics in my comments!

**Pull request merged**